typing/signature_matching.ml

```diff
@@ -0,0 +1,177 @@
+  | 5 | 6 | 7 | 8 -> 2
+  | 9 | 10 | 11 -> 3
+  | len -> len/4
+
+let fuzzy_match_suggestions env compatibility ~subst current =
+  let open Stable_matching in
+  let compatibility = compatibility env subst in
+  let matches =
+    fuzzy_match_names ~max_right_items ~cutoff ~compatibility
+      current.left current.right
+  in
+  match matches.pairs with
+  | [] -> false, subst, current
+  | pairs ->
+      let subst = List.fold_left Suggestion.apply_renaming subst pairs in
+      true, subst, { matches with pairs = pairs @ current.pairs }
+
+type ('a,'b,'c,'d,'v,'cl,'ext) type_kind_map =
+  {
+    module_types: 'a;
+    modules: 'b;
+    types: 'c;
+    class_types: 'd;
+    values: 'v;
+    classes: 'cl;
+    extensions: 'ext
+  }
+
+let empty = {
+  module_types = [];
+  modules = [];
+  types = [];
+  class_types = [];
+  values = [];
+  classes = [];
+  extensions = [];
+}
+
+let classify map = function
+  | Types.Sig_module (_, _, decl, _, _) as item ->
+      { map with modules = Field.make item decl :: map.modules }
+  | Types.Sig_type (_,decl,_,_) as item ->
+      { map with types = Field.make item decl :: map.types }
+  | Types.Sig_modtype (_, decl, _) as item ->
+      let module_types = Field.make item decl :: map.module_types in
+      { map with module_types }
+  | Types.Sig_class_type (_, decl, _, _) as item ->
+      { map with class_types = Field.make item decl :: map.class_types }
+  | Types.Sig_value (_, desc, _) as item ->
+      { map with values = Field.make item desc :: map.values }
+  | Types.Sig_class (_, decl, _, _) as item ->
+      { map with classes = Field.make item decl :: map.classes }
+  | Types.Sig_typext (_,decl,_,_) as item ->
+      { map with extensions = Field.make item decl :: map.extensions }
+
+let init (sgs:Includemod.Error.signature_symptom) =
+  let left = List.fold_left classify empty sgs.additions in
+  let right = List.fold_left classify empty (List.rev sgs.missings) in
+  let init_kind proj =
+    { Stable_matching.left = proj left; pairs = []; right = proj right }
+  in
+  sgs.env, sgs.subst,
+  {
+    module_types = init_kind (fun x -> x.module_types);
+    modules = init_kind (fun x -> x.modules);
+    types = init_kind (fun x -> x.types);
+    class_types = init_kind (fun x -> x.class_types);
+    values = init_kind (fun x -> x.values);
+    classes = init_kind (fun x -> x.classes);
+    extensions = init_kind (fun x -> x.extensions)
+  }
+
+module C = Includemod.Check
+
+let rec iterate env subst lim map =
+  let fuzzy_match c ~subst items = fuzzy_match_suggestions env c ~subst items in
+  let pmty, subst, module_types =
+    fuzzy_match C.module_types ~subst map.module_types in
+  let pm, subst, modules = fuzzy_match C.modules ~subst map.modules in
+  let pty, subst, types = fuzzy_match C.types ~subst map.types in
+  let progress = pmty || pm || pty in
+  if progress && lim > 0 then
+    iterate env subst (lim-1)
+      { map with module_types; modules; types }
+  else subst, map
```

Comment:
I think it might be nice to separate the following logic from this function and the one below:
- a function that does a fold_map over the static components of the map, threading `subst` and returning it alongside `progress`
- a function that does a map over the dynamic/value components of the map
- then the two functions `iterate` and `value_suggestions` defined on top of it

In particular this would make it slightly easier to check that the union of the two helper functions is exhaustive, that all fields of the map are handled as either static or dynamic.

Reply:
Done. I am not fully convinced that the new higher-rank polymorphic functions are easier to review however.